src/drivers/mailbox/bcm2835_mailbox.c

```diff
@@ -0,0 +1,73 @@
+/**
+ * @file
+ * @brief Mailbox Communication mechanism for Raspberry Pi
+ *
+ * @date 16.07.15
+ * @author Michele Di Giorgio
+ */
+
+#include <errno.h>
+#include <stdint.h>
+
+#include <drivers/mailbox/bcm2835_mailbox.h>
+#include <hal/reg.h>
+#include <hal/mem_barriers.h>
+
+static struct bcm2835_mailbox_regs volatile *const mailbox_regs =
+		(struct bcm2835_mailbox_regs volatile *) BCM2835_MAILBOX_BASE;
+
+/**
+ * Read from memory mapped IO registers with a memory barrier around.
+ * @param addr - pointer to register address
+ *
+ * @return data read
+ */
+static uint32_t read_mmio(uint32_t volatile *addr)
+{
+	uint32_t n;
+	data_mem_barrier();
+	n = REG_LOAD(addr);
+	data_mem_barrier();
+	return n;
+}
+
+/**
+ * Write to memory mapped IO register with a memory barrier around.
+ * @param addr - pointer to register address
+ * @param val - data to be stored
+ */
+static void write_mmio(uint32_t volatile *addr, uint32_t val)
+{
+	data_mem_barrier();
+	REG_STORE(addr, val);
+	data_mem_barrier();
```

Comment:
I'm not quite sure I understand you. Following the BCM2835 guide means leaving only a barrier _before_ `REG_STORE` and a barrier _after_ `REG_LOAD`, doesn't it?